decision_service/dsclientlib/ds_driver.cc

```diff
@@ -0,0 +1,131 @@
+#include "ds_driver.h"
+#include "ds_event.h"
+#include "logger/ds_logger.h"
+
+#include <boost/uuid/uuid_generators.hpp> // generators
+#include <boost/uuid/uuid_io.hpp>         // streaming operators etc.
+#include "error_callback_fn.h"
+
+
+//this macro assumes that success_code equals 0
+#define TRY_OR_RETURN(x) do { \
+  int retval = (x); \
+  if (retval != 0) { \
+    return retval; \
+  } \
+} while (0)
```

Comment:
>while (0) [](start = 2, length = 9)

why?

Reply:
Makes macros safe under several scenarios.  See [https://stackoverflow.com/questions/257418/do-while-0-what-is-it-good-for](https://stackoverflow.com/questions/257418/do-while-0-what-is-it-good-for)

---
In reply to: [186734789](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r186734789) [](ancestors = 186734789)

Reply:
Done

---
In reply to: [187047467](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r187047467) [](ancestors = 187047467,186922638,186734789)